packages/site/cms-site/src/iframebridge/IFrameBridge.tsx

```diff
@@ -30,13 +54,82 @@ export const IFrameBridgeContext = React.createContext<IFrameBridgeContext>({
     sendMessage: () => {
         //empty
     },
+    previewElementsData: [],
+    removePreviewElement: () => {
+        // empty
+    },
+    addPreviewElement: () => {
+        // empty
+    },
 });
 
 export const IFrameBridgeProvider: React.FunctionComponent = ({ children }) => {
     const [block, setBlock] = React.useState<unknown | undefined>(undefined);
     const [selectedAdminRoute, setSelectedAdminRoute] = React.useState<string | undefined>(undefined);
     const [hoveredAdminRoute, setHoveredAdminRoute] = React.useState<string | undefined>(undefined);
     const [showOutlines, setShowOutlines] = React.useState<boolean>(false);
+    const [previewElements, setPreviewElements] = React.useState<PreviewElement[]>([]);
+    const [recalculatePreviewDataIndex, setRecalculatePreviewDataIndex] = React.useState<number>(0);
+
+    const childrenWrapperRef = React.useRef<HTMLDivElement>(null);
+
+    const triggerRecalculationOfPreviewData = React.useCallback(() => {
+        setRecalculatePreviewDataIndex((index) => index + 1);
+    }, []);
+
+    React.useEffect(() => {
+        const mutationObserver = new MutationObserver(() => {
+            triggerRecalculationOfPreviewData();
+        });
+
+        const resizeObserver = new ResizeObserver(() => {
+            triggerRecalculationOfPreviewData();
+        });
+
+        if (childrenWrapperRef.current) {
```

Comment:
Shouldn't the observers only be instantiated if the ref is available? Also, they are currently disconnected in all cases, even if they don't observe in the first place.